be more explicit about callback chain

jdamiba opened the pull request:
The purpose of this PR is to add a section to the `Advanced Callbacks` page which describes the callback chain in more detail. 

closes https://github.com/plotly/dash-docs/issues/852
closes https://github.com/plotly/dash-docs/issues/755
closes https://github.com/plotly/dash-docs/issues/874

jdamiba commented:
@alexcjohnson Can you please review this?

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.py:
```diff
@@ -89,5 +89,21 @@ def slow_function(input):
+    rc.Markdown('''
+    ## The Callback Chain
+
+    - Upon initial load of a Dash app or based on later interactions, all of the related callbacks are collected, based on both the existence of newly-created outputs and newly-changed inputs.
```
I think we need some examples or other clarification for a number of these points. Ideally I think it'd be nice to have diagrams showing how callbacks can chain, and at least one of these diagrams should be presented alongside explicit dash code - kind of like the callback graphs in the devtools but probably simplified.

> all of the related callbacks are collected

Make it clear that this is recursive: we collect not only those whose inputs changed directly, but also those whose inputs include outputs of another callback we've already collected. It's important that we collect the entire chain up front, or we wouldn't be able to tell for sure which callbacks are blocking others. For example: dropdown A is an input for one callback (X) that sets the value of dropdown B. A second callback (Y) has both dropdown A and B as inputs and outputs to store C. A third callback (Z) uses the store to create a graph D. All three callbacks are collected, but Y is blocked until X completes, and Z is blocked until Y completes.

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.py:
```diff
@@ -89,5 +89,21 @@ def slow_function(input):
+    ## The Callback Chain
+
+    - Upon initial load of a Dash app or based on later interactions, all of the related callbacks are collected, based on both the existence of newly-created outputs and newly-changed inputs.
+      - Upon initial load or in the case of callbacks returning `children` with new compononets, none of the inputs are conisidered to have "changed" unless (1) they're outputs of another callback, or (2) the callback is updating something *outside* the new `children` (not possible upon initial load).
```
> none of the inputs are conisidered to have "changed" unless

"changed" is not only relevant to the logic here, it also determines what the user sees in `callback_context.triggered` - let's point that out and link to the `callback_context` docs.

> (1) they're outputs of another callback

If we walk through an example like the three-callback chain I describe above, we can show exactly what `callback_context.triggered` looks like inside each callback.

> (2) the callback is updating something *outside* the new `children` (not possible upon initial load)

This is a pretty subtle point that probably needs an example all to itself.

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.py:
```diff
@@ -89,5 +89,21 @@ def slow_function(input):
+
+    - Upon initial load of a Dash app or based on later interactions, all of the related callbacks are collected, based on both the existence of newly-created outputs and newly-changed inputs.
+      - Upon initial load or in the case of callbacks returning `children` with new compononets, none of the inputs are conisidered to have "changed" unless (1) they're outputs of another callback, or (2) the callback is updating something *outside* the new `children` (not possible upon initial load).
+      - But, the callback will still be queued as an "initial call", unless ALL of its outputs are outputs of something else- then it's no longer considered an "initial call", but it's still put into the callback queue on the assumption that those outputs will change.
```
We should mention `prevent_initial_call` here - #874 (may want to describe that in more than one place, or maybe here is sufficient?)

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.py:
```diff
@@ -89,5 +89,21 @@ def slow_function(input):
+    - Upon initial load of a Dash app or based on later interactions, all of the related callbacks are collected, based on both the existence of newly-created outputs and newly-changed inputs.
+      - Upon initial load or in the case of callbacks returning `children` with new compononets, none of the inputs are conisidered to have "changed" unless (1) they're outputs of another callback, or (2) the callback is updating something *outside* the new `children` (not possible upon initial load).
+      - But, the callback will still be queued as an "initial call", unless ALL of its outputs are outputs of something else- then it's no longer considered an "initial call", but it's still put into the callback queue on the assumption that those outputs will change.
+    - Callbacks are then executed in the order that their inputs are ready.
```
We should note that many server-side callbacks can be requested simultaneously, if they're all unblocked together - and assuming you're running a multi-process server they can execute in parallel.

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.py:
```diff
@@ -89,5 +89,21 @@ def slow_function(input):
+      - But, the callback will still be queued as an "initial call", unless ALL of its outputs are outputs of something else- then it's no longer considered an "initial call", but it's still put into the callback queue on the assumption that those outputs will change.
+    - Callbacks are then executed in the order that their inputs are ready.
+      - First dispatched are callbacks that have no inputs that are the outputs of another callback.
+      - As callbacks return, if their outputs were prevented from updateing, these are removed as "changed" in any other callbacks that they are inputs for. Either way, they're marked as no longer blocking the other callback.
```
> prevented from updateing

using `raise PreventUpdate` or `return dash.no_update`

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.py:
```diff
@@ -89,5 +89,21 @@ def slow_function(input):
+    - Callbacks are then executed in the order that their inputs are ready.
+      - First dispatched are callbacks that have no inputs that are the outputs of another callback.
+      - As callbacks return, if their outputs were prevented from updateing, these are removed as "changed" in any other callbacks that they are inputs for. Either way, they're marked as no longer blocking the other callback.
+      - If a callback has no changed props and is not an "initial call", then it is removed from the callback queue.
```
And all of its outputs are also removed as "changed" in other callbacks, possibly resulting in these callbacks being removed as well.

nicolaskruchten commented:
Notes:

* I recommend addressing #755 and #874 in this same PR, so you can refer to those new sections from the callback-chain stuff
* I recommend stressing under what circumstances the front-end makes distinct HTTP calls the back-end (i.e. once per individual callback firing) and under what circumstances the front-end will issue simultaneous HTTP calls (i.e. once per callback for which the front-end doesn't already know the result will be invalidate by a call it already knows it's got to do)

jdamiba commented:
@alexcjohnson Can you take another look at this PR? I've reorganized what I had written at `/advanced-callbacks` with some example apps and added a section about the Dash app lifecycle at `/app-lifecycle`.

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.R:
```diff
@@ -94,10 +94,26 @@ The [Performance](/performance) section of the Dash docs delves a
+
+    - Upon initial load of a Dash app or based on later interactions, all of the related callbacks are collected, based on both the existence of newly-created outputs and newly-changed inputs.
+      - Upon initial load or in the case of callbacks returning `children` with new compononets, none of the inputs are conisidered to have "changed" unless (1) they're outputs of another callback, or (2) the callback is updating something *outside* the new `children` (not possible upon initial load).
+      - But, the callback will still be queued as an "initial call", unless ALL of its outputs are outputs of something else- then it's no longer considered an "initial call", but it's still put into the callback queue on the assumption that those outputs will change.
```
```suggestion
      - But, the callback will still be queued as an "initial call", unless ALL of its inputs are outputs of something else- then it's no longer considered an "initial call", but it's still put into the callback queue on the assumption that those inputs will change.
```

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.R:
```diff
@@ -94,10 +94,26 @@ The [Performance](/performance) section of the Dash docs delves a
+      - But, the callback will still be queued as an "initial call", unless ALL of its outputs are outputs of something else- then it's no longer considered an "initial call", but it's still put into the callback queue on the assumption that those outputs will change.
+    - Callbacks are then executed in the order that their inputs are ready.
+      - First dispatched are callbacks that have no inputs that are the outputs of another callback.
+      - As callbacks return, if their outputs were prevented from updateing, these are removed as "changed" in any other callbacks that they are inputs for. Either way, they're marked as no longer blocking the other callback.
```
```suggestion
      - As callbacks return, if their outputs were prevented from updating, these are removed as "changed" in any other callbacks that they are inputs for. Either way, they're marked as no longer blocking the other callback.
```

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.py:
```diff
@@ -89,5 +89,193 @@ def slow_function(input):
+
+    It is important to note that when a Dash app is initially loaded in a web browser by the `dash-renderer` front-end client, its entire callback chain is introspected recursively.
+
+    This allows the `dash-renderer` to predict the order in which callbacks will need to be executed, as callbacks are blocked when their inputs are outputs of other callbacks which have not yet fired. In order to unblock the execution of these callbacks, first callbacks whose inputs are immediately available must be executed. This process helps the `dash-renderer` to minimize the number of netwo ...
```
```suggestion
    This allows the `dash-renderer` to predict the order in which callbacks will need to be executed, as callbacks are blocked when their inputs are outputs of other callbacks which have not yet fired. In order to unblock the execution of these callbacks, first callbacks whose inputs are immediately available must be executed. This process helps the `dash-renderer` to minimize the time and effort it uses, and avoid unnecessarily redrawing the page, by making sure it only requests that a callback is executed when all of the callback's inputs have reached their final values.
```

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.py:
```diff
@@ -89,5 +89,193 @@ def slow_function(input):
+    [Input("button", "n_clicks")]
+)
+def change_text(n_clicks):
+    if n_clicks is None:
+        return ["n_clicks is None", "n_clicks is None"]
```
```suggestion
```

`str(None)` is `"None"` so no special handling is needed

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.py:
```diff
@@ -89,5 +89,193 @@ def slow_function(input):
+    Examine the following Dash app:
+    '''),
+
+    rc.Syntax('''
```
Can we do `rc.Syntax` / `rc.Example` pairs with these new examples, like we do most other places, so users can see the apps in action alongside the code?

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.py:
```diff
@@ -89,5 +89,193 @@ def slow_function(input):
+    '''),
+
+    rc.Markdown('''
+    Notice that when this app is finished being loaded by a web browser and ready for user interaction, the `children` of the `html.Div` components is not the text that is defined when the components are declared in the app's layout but rather is the result of the `change_text()` callback being executed. This is because the "initial call" of the callback occured with `n_clicks` having the value o ...
```
```suggestion
    Notice that when this app is finished being loaded by a web browser and ready for user interaction, the `html.Div` components do not say "callback not executed" as declared in the app's layout, but rather "n_clicks is None," the result of the `change_text()` callback being executed. This is because the "initial call" of the callback occurred with `n_clicks` having the value of `None`.
```

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.py:
```diff
@@ -89,5 +89,193 @@ def slow_function(input):
+
+    Most frequently, callbacks are executed as a direct result of user interaction, such as clicking a button or selecting an item in a dropdown menu. When such interactions occur, Dash components communicate their new values to the `dash-renderer` front-end client, which then requests that the Dash server execute any callback function that has the newly changed value as input.
+
+    If a Dash app has multiple callbacks, the `dash-renderer` batches requests for callbacks to be executed based on whether or not they can be immediately executed with the newly changed inputs. If several inputs change simultaneously, then requests are made to execute them all.
```
```suggestion
    If a Dash app has multiple callbacks, the `dash-renderer` requests callbacks to be executed based on whether or not they can be immediately executed with the newly changed inputs. If several inputs change simultaneously, then requests are made to execute them all.
```

I wouldn't want to give the impression that we merge multiple callbacks into a single request, which "batches requests" might imply. Someday we might do this, but not today :)

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.py:
```diff
@@ -89,5 +89,193 @@ def slow_function(input):
+    rc.Markdown('''
+    ## Prevent Callbacks From Being Executed on Initial Load
+
+    You can use the `prevent_initial_call` attribute of callbacks to prevent callbacks from being fired when the app initially loads, as in the following example app.
```
I don't see `prevent_initial_call` below?

alexcjohnson commented on dash_docs/chapters/app_lifecycle/index.py:
```diff
@@ -0,0 +1,27 @@
+examples = tools.load_examples(__file__)
+
+layout = html.Div(children=[
+    rc.Markdown('''# Dash App Life Cycle
```
```suggestion
    rc.Markdown('''
    # Dash App Life Cycle
```
To ensure our automatic dedenting of markdown works correctly

alexcjohnson commented on dash_docs/chapters/app_lifecycle/index.py:
```diff
@@ -0,0 +1,27 @@
+
+    This section describes the lifecyle of a Dash app.
+
+    1. When `python app.py` or `gunicorn app:server` is run, all of the files in a Dash app are executed. This means that if you have a statement such as `df = pd.read_csv('...')`, it is run when the program starts, rather than when the page is loaded. Therefore, if the CSV changes after the program starts, the `df` will not be updated until the program restarts. For this reason, it is recommende ...
```
```suggestion
    1. When `python app.py` or `gunicorn app:server` is run, all of the files in a Dash app are executed. This means that if you have a statement such as `df = pd.read_csv('...')`, it is run when the program starts, rather than when the page is loaded. Therefore, if the CSV changes after the program starts, `df` will not be updated until the program restarts. In this case, it is recommended to provide data via a periodic `celery` task or by setting `app.layout` as a function to regenerate the layout with each page load.
```

alexcjohnson commented on dash_docs/chapters/app_lifecycle/index.py:
```diff
@@ -0,0 +1,27 @@
+        - Note that this is a recursive process:  the `dash-renderer` collects not only those callbacks whose inputs can be changed directly through user interaction, but also those callbacks whose inputs are outputs of another callback that has already been collected. It's important that the `dash-renderer` collects the entire callback chain up front, or else it wouldn't be able to determine whi ...
+
+    3. All callbacks that have inputs in currently the app layout are executed. This is known as the "initial call" of the callbacks, and this behavior can be suppressed by using the `prevent_initial_call` attribute.
+        - If a Dash app has multiple callbacks, the `dash-renderer` batches requests for callbacks to be executed based on whether or not they can be immediately executed with the newly changed inputs.
```
```suggestion
        - If a Dash app has multiple callbacks, the `dash-renderer` requests callbacks to be executed based on whether or not their inputs might be changed as a result of another callback.
```

jdamiba commented:
@alexcjohnson I have updated this PR with your suggested commits as well and used the `rc.Syntax/rc.Example` workflow to interactively embed the example Dash apps. 

Please note that while I have updated the text in the R `Advanced Callbacks` chapter to match the Python text, I have not yet created R versions of the example apps or created an R version of the `Dash App Lifecycle` chapter.

alexcjohnson commented:
> I have updated this PR with your suggested commits as well and used the rc.Syntax/rc.Example workflow

Fantastic, love it!

> Please note that while I have updated the text in the R Advanced Callbacks chapter to match the Python text, I have not yet created R versions of the example apps or created an R version of the Dash App Lifecycle chapter.

OK great, glad that's on your radar. It's looking really good, but I'll hold off a final review until the rest of the R pieces are in place.

Not to be implemented in this PR, but I noticed this and started wondering if we should try to DRY some of this across languages -> #950

jdamiba commented:
@alexcjohnson This PR is ready for your review- I have made R versions of the example apps and put them in the appropriate place, and have created a "Dash App Lifecycle" chapter for the R version as well. 

Please note that since https://github.com/plotly/dashR/issues/227 is outstanding, I cannot document that feature on the R side.

alexcjohnson commented on dash_docs/chapters/advanced_callbacks/index.py:
```diff
@@ -89,5 +89,72 @@ def slow_function(input):
+    rc.Markdown('''
+    ## When Are Callbacks Executed?
+
+    This section describes the circumstances under which the `dash-renderer` front-end client can make a request to the Dash back-end server to execute a callback function.
```
```suggestion
    This section describes the circumstances under which the `dash-renderer` front-end client can make a request to the Dash back-end server (or the clientside callback code) to execute a callback function.
```
just to make it clear that all of this applies exactly the same way to clientside callbacks.

alexcjohnson commented on dash_docs/chapters/app_lifecycle/index.py:
```diff
@@ -0,0 +1,28 @@
+
+    1. When `python app.py` or `gunicorn app:server` is run, all of the files in a Dash app are executed. This means that if you have a statement such as `df = pd.read_csv('...')`, it is run when the program starts, rather than when the page is loaded. Therefore, if the CSV changes after the program starts, `df` will not be updated until the program restarts. In this case, it is recommended to pr ...
+
+    2. The `app.layout` is rendered and the app's callback chain is introspected to determine the relationships between inputs, callbacks, and outputs.
```
Step 1 is happening entirely on the server, but step 2 is in the browser without describing how we got there. Maybe we want a step between like:

2 - When the page loads and `dash-renderer` starts, it asks the server for both the initial layout (`app.layout`) and a list of all the callbacks registered to the app. This list of callbacks is only provided at page load and can't be modified later, so the app needs to know about all of its callbacks from the beginning even for components that do not exist in the initial layout.

alexcjohnson reviewed (approved):
@jdamiba this looks great! I just made 2 small comments (commented on the Python side but they apply to both Py and R) and then I think this is ready! Nice job on the `prevent_initial_call` part, and you handled the Py/R difference there well.
